### ISSUE TITLE
Bug: strict json_schema decode wedges (no token progress, >5min) on Qwen3-Coder-30B-A3B

### ISSUE THREAD

**Issue by vinayvobbili** [waybarrios/vllm-mlx#546]
### Environment

- `vllm-mlx 0.3.0` (pip)
- Mac Studio (Apple Silicon, 96 GB)
- Model: `Qwen3-Coder-30B-A3B-Instruct-8bit` (sparse MoE, 3B active per token)
- Launch args:
  ```
  vllm_mlx_server.py \
    --tool-call-parser qwen3_coder \
    --model Qwen3-Coder-30B-A3B-Instruct-8bit \
    --port 8003 --host 0.0.0.0 \
    --served-model-name qwen3-coder-30b-a3b \
    --max-request-tokens 131072 \
    --default-temperature 0.3
  ```
- Plus our local multi-slot LRU patch (follow-up to #523). Happy to retest against stock to confirm independence.

### Symptom

A `/v1/chat/completions` request with `response_format={"type":"json_schema","json_schema":{...},"strict":true}` for a moderately complex Pydantic-derived schema gets stuck mid-decode for 5+ minutes. `[disconnect_guard]` polls keep counting up while no token-progress logs appear. The request eventually 504s via the upstream proxy. Every subsequent request queued behind it also fails to progress until the server is restarted.

### Schema (the one that wedged)

```json
{
  "title": "NoveltyLLMResponse",
  "type": "object",
  "$defs": {
    "VulnerableProductMention": {
      "type": "object",
      "properties": {
        "product": {"type": "string"},
        "vendor": {"anyOf": [{"type":"string"},{"type":"null"}]},
        "version_constraint": {"anyOf": [{"type":"string"},{"type":"null"}]}
      },
      "required": ["product","vendor","version_constraint"],
      "additionalProperties": false
    }
  },
  "properties": {
    "novelty_score": {"type":"integer","minimum":1,"maximum":10},
    "novelty_label": {"type":"string","enum":["Seen Before","Familiar","Mostly New","Net New"]},
    "summary": {"type":"string"},
    "recommendation": {"type":"string"},
    "whats_new_reasons": {"type":"array","items":{"type":"string"}},
    "whats_familiar_reasons": {"type":"array","items":{"type":"string"}},
    "vulnerable_products": {"type":"array","items":{"$ref":"#/$defs/VulnerableProductMention"}}
  },
  "required": ["novelty_score","novelty_label","summary","recommendation","whats_new_reasons","whats_familiar_reasons","vulnerable_products"],
  "additionalProperties": false
}
```

Likely hot spots: `anyOf [string,null]` (from `Optional[str]`), enum + integer min/max, nested `$defs`, `additionalProperties:false`. Prompt was ~12.6K chars (system+user), temp 0.1, `max_tokens=null`, no tools.

### Log signature

```
INFO:vllm_mlx.server:[REQUEST] POST /v1/chat/completions stream=False model='qwen3-coder-30b-a3b' max_tokens=None temp=0.1 ... response_format=type='json_schema' ...
INFO:vllm_mlx.server:[disconnect_guard] poll #100 disconnected=False elapsed=50.8s
INFO:vllm_mlx.server:[disconnect_guard] poll #200 disconnected=False elapsed=101.7s
...
INFO:vllm_mlx.server:[disconnect_guard] poll #590 disconnected=False elapsed=299.5s
"POST /v1/chat/completions HTTP/1.1" 504 Gateway Timeout
```

No `JSON validation failed`, no exception traceback — the decoder is simply not producing tokens.

### Workaround

Switching the client to `response_format={"type":"json_object"}` (no schema; Pydantic validation post-hoc) bypasses the wedge entirely. We're now running this in prod (LangChain `with_structured_output(schema, method="json_mode")`).

### What seems specific to Qwen3-Coder-30B-A3B

The same client schema + prompt against `GLM-4.7-Flash-8bit` (dense, same vllm-mlx version) ran fine for months. The wedge surfaced only after we swapped the analysis backend GLM → Qwen3-Coder. Hypothesis: sparse-MoE logit distributions can concentrate probability mass on tokens disallowed by the FSM in a way that creates a sampling loop the enforcer's recovery path doesn't break out of — but I haven't isolated the sampler-side livelock yet.

Not directly the same as #538 (which addresses HTTP 200 + invalid content); this is "decoder never produces tokens."

cc @waybarrios @Thump604 @janhilgard — would value a sanity check on whether the lm-format-enforcer integration has known issues with sparse-MoE Coder models, or whether this points at a sampler-side livelock.

**Comment by Thump604** [comment]
I opened #547 as a narrow constrained-decoding hardening patch.

What I checked:
- current `waybarrios/vllm-mlx` main at `7e30484`
- the `v0.3.0` tag mentioned in the report
- `JSONSchemaLogitsProcessor` construction and per-step calls with the schema from this issue and `Qwen/Qwen3-Coder-30B-A3B-Instruct` tokenizer data

Those processor-only checks did not hang on schema construction or individual processor calls. The concrete issue I could verify is that the allowed-token set permits unbounded pure JSON whitespace at legal positions such as the root and after `{`, which can make a non-streaming request appear to produce no useful JSON progress while it holds the slot.

#547 bounds that non-progress whitespace path: normal whitespace remains allowed, but after a long trailing whitespace run outside a JSON string, pure-whitespace tokens are masked so the next token must make structural/content progress.

What I am not claiming: I am not calling this a proven sparse-MoE root cause, and I am not claiming it fully explains every possible 5-minute stall. It is a verified narrow decoder-progress gap with tests. CI is green on #547.

**Comment by vinayvobbili** [comment]
Thanks @Thump604 — appreciate the careful scoping and the honest "not the proven root cause" framing.

The whitespace-progress gap matches the symptom we saw (slot held, `[disconnect_guard]` polling, zero token-progress logs), even if it isn't necessarily the only path that can wedge a Coder request. The fact that processor-only checks against `v0.3.0` and current main don't hang is a useful data point on its own — narrows what's left to investigate if a future wedge shows up after #547 lands.

Plan on our side:

1. **Keep the prod workaround** (`response_format={"type":"json_object"}` + post-hoc Pydantic validation via LangChain `with_structured_output(..., method="json_mode")`). It's been stable for the analysis backend and the strict-schema path has no callers right now.
2. **Re-test strict `json_schema` once #547 ships** in a tagged release. I'll rerun the same client (NoveltyLLMResponse schema, ~12.6K char prompt, Qwen3-Coder-30B-A3B-Instruct-8bit, temp 0.1) and post results back here.
3. If the wedge reproduces after the patch, I'll keep the issue open and try to capture a sampler-side trace (token-id stream around the stall point) so we have something concrete on the lm-format-enforcer / MoE-logits interaction angle rather than circumstantial timing.

Happy to leave #546 open until step 2 lands so we don't lose the thread if there's a second failure mode hiding behind this one.

**Comment by Thump604** [comment]
I dug one layer deeper after this comment. My current read is that #547 is useful hardening, but I would not treat pure-whitespace progress as the full root-cause explanation yet.

What I checked locally:

- `waybarrios/vllm-mlx` #547 head `92acdab3056fed3b5c0f3443edc0a28e99374843`
- the issue schema from this thread
- the `Qwen/Qwen3-Coder-30B-A3B-Instruct` tokenizer snapshot available locally
- `JSONSchemaLogitsProcessor` / lm-format-enforcer allowed-token behavior with sequential prefix-state, not just direct final-prefix calls
- the non-stream `/v1/chat/completions` path through `server._wait_with_disconnect()` and `SimpleEngine._stream_generate_text()`

The important correction: with sequential LMFE prefix state and the Qwen3-Coder tokenizer, long pure-whitespace runs do get forced back to structural/content tokens by the enforcer state itself. For this schema I saw:

- at root: 282 allowed tokens, 265 pure whitespace tokens
- after one root space: 269 allowed, 252 pure whitespace
- after 256 root spaces: 7 allowed, 0 pure whitespace
- after `{`: 269 allowed, 265 pure whitespace
- after `{ `: 256 allowed, 252 pure whitespace
- after `{` plus 256 spaces: 3 allowed, 0 pure whitespace

So #547 still closes a real generic progress gap and is safe, but if this exact wedge reproduces after #547, I would look next at request/engine behavior rather than assuming sparse-MoE + whitespace is proven.

The code path has two details that match the observed log signature:

1. The launch in the report sets `--max-request-tokens 131072` but not `--max-tokens`. In current server code, `max_tokens=null` resolves to the server default `--max-tokens` value, which is 32768. The default request timeout is 300s. So a non-stream strict-schema request can spend the whole 300s generating or attempting to generate before any response body is visible to the client.

2. For non-stream chat, `_wait_with_disconnect()` awaits the whole `engine.chat(...)` coroutine. The "no token-progress logs" line is therefore not proof that no tokens were generated; the server only logs the chat completion token count after `engine.chat()` returns. On the SimpleEngine text route, cancellation is cooperative around `mlx_lm.stream_generate()` yields: `abort_event` is checked after a response chunk is yielded. If strict decoding wedges before the next yield, or is simply very slow before the first/next token, the worker can keep holding the generation lock after the proxy/client disconnects. That explains the "subsequent requests queue until restart" symptom without yet proving the decoder-level cause.

Concrete next repro split I would use:

- First retest on stock vllm-mlx without the local LRU patch, to remove that variable.
- Run the same schema with `stream=true`. Streaming should use the same constrained processor, but it reveals whether tokens/chunks are actually being produced. If chunks arrive, capture the last 20 token ids / decoded repr before the slow point. If no first chunk arrives, the stall is before the first token yield.
- Run non-stream with explicit small `max_tokens` values, e.g. 512 and 1024, and a timeout longer than expected. If it returns `finish_reason=length`, this is long constrained generation rather than a hard no-progress deadlock.
- Try adding schema bounds for the free-form fields (`maxLength` on strings and `maxItems` on arrays) as a diagnostic. The posted schema has several unbounded string/array surfaces, so strict schema decoding still leaves very large valid continuations available.
- If the wedge persists, instrument `JSONSchemaLogitsProcessor.__call__` for this request with: suffix token length, decoded suffix tail repr, allowed-token count, pure-whitespace count, JSON context, disabled state, and whether this is before first emitted token. That should distinguish: whitespace progress, valid-but-long JSON generation, empty-allowed-set recovery, or cancellation/lock retention.

What I am not claiming: I still have not reproduced the full model-side wedge with the exact MLX model weights, and I am not calling sparse-MoE the root cause. The strongest proven points so far are that processor construction/per-step calls do not hang in isolation, pure whitespace alone is probably incomplete as a root-cause theory for this schema/tokenizer, and the non-stream SimpleEngine path currently lacks enough token-boundary visibility to tell long generation from a pre-yield decoder stall.

**Comment by Thump604** [comment]
I opened #548 with the narrower engine-path fix that came out of this triage.

What it changes:
- non-stream LLM chat with request-local `logits_processors` now uses the existing stream aggregation path
- ordinary non-stream chat without constrained decoding stays on `model.chat()`
- regression coverage proves constrained-decoding kwargs reach `stream_generate` and do not call the blocking chat API

Why this is relevant to this report: `response_format` is implemented through request-local logits processors. On the SimpleEngine LLM path, the previous non-stream branch could sit inside blocking `model.chat()` until completion, which meant no token-boundary visibility or cooperative cancellation while the serialized generation lock was held.

Validation on the PR:
- focused red/green regression
- `tests/test_simple_engine.py`, `tests/test_constrained_decoding.py`, and `tests/test_structured_output.py`: `112 passed, 2 skipped`
- ruff, black check, `git diff --check`
- Runtime upstream-local-repro preflight against clean upstream main

What I am not claiming: this does not prove sparse-MoE behavior is the cause, and I still did not run the exact reporter MLX weights locally. This fixes the local engine-path defect that matched the reported non-stream wedge/queued-requests shape.

**Comment by vinayvobbili** [comment]
Thanks @Thump604 — #548 is the one that lines up with what we saw in prod.

The signature in the report that I think pins this on the engine path rather than a sampler livelock is the *queued-requests* part: not just the first request stalling, but every subsequent `/v1/chat/completions` failing to progress until the server was restarted. A whitespace livelock would (mostly) stall the offending request and let other requests through once the slot freed; a serialized-lock-held-inside-blocking-`model.chat()` matches the "everyone behind it stops" shape exactly. So #548's framing fits the symptom signature better than #547 does, even with the same "not the proven sparse-MoE cause" hedge.

Updated plan on our side:

1. **Workaround stays** (`json_object` + post-hoc Pydantic) until both #547 and #548 land in a tagged release.
2. **Re-test once a release ships them.** Same client (NoveltyLLMResponse strict schema, ~12.6K char prompt, `Qwen3-Coder-30B-A3B-Instruct-8bit`, temp 0.1) plus an explicit *queued-requests* check (fire a second request 30s into a blocking one and confirm it progresses) — that's the half of the original symptom #548 should fix outright.
3. If a wedge still reproduces after both patches land, I'll capture a sampler-side trace (token-id stream around the stall point) before commenting again, instead of just reporting "still hangs."

Happy to leave #546 open until step 2 lands so we don't lose the thread.
